PhabricatorBMOAuthProvider.php

```diff
@@ -0,0 +1,400 @@
+      $errors[] = pht('Application Name is required.');
+      $issues[self::CONFIG_KEY_APP_NAME] = pht('Required');
+    }
+
+    $trans_code_length =
+      (int) self::getArrayValueOrDefault($values, self::CONFIG_KEY_TRANSACTION_CODE_LENGTH);
+    if($trans_code_length < 16 || $trans_code_length > 64) {
+      $errors[] = pht('Transaction Code Length must be between 16 and 64.');
+      $issues[self::CONFIG_KEY_TRANSACTION_CODE_LENGTH] = pht('Invalid');
+    }
+
+    // If there are no errors, set the config's providerDomain column
+    if(!count($errors)) {
+      $config = $this->getProviderConfig();
+      $config->setProviderDomain($values[self::CONFIG_KEY_BUGZILLA_DOMAIN]);
+    }
+
+    return array($errors, $issues, $values);
+  }
+
+  public function getAdapter() {
+    if (!$this->adapter) {
+      $this->adapter = $adapter = new PhabricatorBMOAuthAdapter();
+      $this->configureAdapter($adapter);
+    }
+    return $this->adapter;
+  }
+
+  public function getConfig() {
+    if (!$this->providerConfig) {
+      $this->providerConfig = $this->getProviderConfig();
+    }
+    return $this->providerConfig;
+  }
+
+  public function configureAdapter($adapter) {
+    $config = $this->getConfig();
+
+    $adapter->setAdapterType(self::ADAPTER_TYPE);
+    $adapter->setAdapterDomain($config->getProperty(self::CONFIG_KEY_BUGZILLA_DOMAIN));
+    $adapter->setAuthenticateURI(
+      id(new PhutilURI(''))
+        ->setProtocol(
+          $config->getProperty(self::CONFIG_KEY_BUGZILLA_PROTOCOL)
+        )
+        ->setDomain($adapter->getAdapterDomain())
+        ->setPath('/auth.cgi')
+        ->setQueryParam('description', $config->getProperty(self::CONFIG_KEY_APP_NAME))
+    );
+  }
+
+  /*
+   *  Methods to guide the user through the auth and login processes
+   */
+
+  protected function renderLoginForm(AphrontRequest $request, $mode) {
+    $adapter = $this->getAdapter();
+
+    $csrf = $this->getAuthCSRFCode($request);
+    $login_uri = $this->getLoginURI();
+    $uri = new PhutilURI($adapter->getAuthenticateURI());
+
+    $uri->setQueryParam('callback',
+      PhabricatorEnv::getURI($login_uri).$csrf.'/');
+
+    $attributes = array('method' => 'GET', 'uri' => (string) $uri);
+    return $this->renderStandardLoginButton($request, $mode, $attributes);
+  }
+
+  public function processLoginRequest(
+    PhabricatorAuthLoginController $controller) {
+
+    $request = $controller->getRequest();
+
+    // When Bugzilla sends us a "back channel" POST during registration...
+    if ($request->isHTTPPost()) {
+      return $this->processLoginRequestBackChannelPost($controller, $request);
+    }
+
+    // The last step in Bugzilla's verification process
+    return $this->processLoginRequestConfirmationGet($controller, $request);
+  }
+
+  private function processLoginRequestBackChannelPost($controller) {
+    $config = $this->getConfig();
+    $account = null;
+    $response = null;
+
```

Comment:
Upon more thought, it may just be another hoop to jump through.  If someone is cURL'ing, they can easily spoof it.